### ISSUE TITLE
get CLI to run synchronously for chainability

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#130]
If there's any way to configure `commander` to be synchronous, and wait for the callback, that'd be good so that the commandline version is chainable -- like:

`sequencer -i photo.jpg && mv photo.jpg ~/Desktop/`

https://npmjs.org/package/commander

**Comment by tech4GT** [comment]
I would like to work on this issue

**Comment by jywarren** [comment]
That'd be super, thank you!

**Comment by tech4GT** [comment]
@jywarren do we want to draw progress bar and loading icon?

**Comment by jywarren** [comment]
That sounds really cool! Let's open a separate issue for it though, so we
can solve things modularly;
https://publiclab.org/notes/warren/01-18-2018/software-outreach-modularity-is-great-for-collaboration

That'd be a perfect next step after this one. Great idea!

**Comment by tech4GT** [comment]
Amazing read @jywarren

**Comment by tech4GT** [comment]
@jywarren should i open the issue?

**Comment by jywarren** [comment]
Yes please! Thanks!!

**Comment by tech4GT** [comment]
@jywarren can you please elaborate on this issue a little. If we store the output image in the same directory, wouldn't that make it chainable too?

**Comment by jywarren** [comment]
What i've seen is that if you run `sequencer -i photo.jpg && mv photo.jpg ~/Desktop/` and the initial `sequencer` command takes any time at all, the `mv` command will run before `sequencer` is done, since it's asynchronous. So I just want to suggest we make it synchronous -- only return complete once the output file is ready -- so that it can be used in shell script command chaining. Does that make sense?

**Comment by tech4GT** [comment]
@jywarren Cool, that makes perfect sense, I will start looking for a workaround now.
Thankyou

**Comment by jywarren** [comment]
:-) thank you!!!

**Comment by tech4GT** [comment]
@jywarren i think this is happening because our node process exits after calling all the async code, so we need to stop the node process from finishing before the images are exported, we can use es6 features for that.
If we declare exportbin function as async and await it in index.js then we can exit the process after the images are exported
Please guide me if my thinking is correct?

**Comment by jywarren** [comment]
Sure, i think that sounds good. I thought there was a way to run a function
synchronously somehow, but this sounds just as good. Thanks!

**Comment by tech4GT** [comment]
@jywarren we would have to change quite a bit of code to es6 i hope it's fine?😅

**Comment by jywarren** [comment]
would that affect browser support?

**Comment by tech4GT** [comment]
It will not work with internet explorer, will that be ok?

**Comment by jywarren** [comment]
I wonder -- based on what we use already right now, what version of IE do
we support?

**Comment by tech4GT** [comment]
<img width="1273" alt="screen shot 2018-02-09 at 4 17 03 pm" src="https://user-images.githubusercontent.com/25617855/36024126-bbf6651e-0db4-11e8-80b3-3b9b0611820e.png">

**Comment by jywarren** [comment]
This sounds great then!

**Comment by tech4GT** [comment]
@jywarren I think there was some error earlier, i just tried chaining commands and it worked. I tried it with edge detection which takes some time to process and it still worked. I moved the output of the sequencer to desktop and it worked😅
`node ./index.js -i ../imgs/myImage.png -s "edge-detect" && mv ./output/sequencer101/image1_1.png ../`

**Comment by jywarren** [comment]
Oh, so it's already chainable then?????!!

**Comment by tech4GT** [comment]
Yes if we use a double &😁

**Comment by jywarren** [comment]
Ah, that makes sense! Let's add that to the docs. Great!

**Comment by tech4GT** [comment]
Sure I'm opening another issue for that...this one can be closed😁

**Comment by tech4GT** [comment]
But the initial example that you took won't work because of poor cli syntax, i have put it in my proposal in detail
https://publiclab.org/notes/tech4gt/02-19-2018/gsoc-proposal-image-sequencer-v2-processing-on-steroids

**Comment by jywarren** [comment]
Sorry, can you explain that here? Just copy in your explanation? 

> does not support naming of the output image

is the key, right?

**Comment by tech4GT** [comment]
yeah sure,
Yes you are right that's the key and we also export the images in a "sequencer + <no. of iterations of sequencer>" folder which makes it difficult to access the output on the fly, also i am planning to implement extension detection which will make it possible to name output of a .jpg image  as .png and it will be automatically exported as png. For eg
sequencer -i "photo.jpeg" -s "invert" -n "photo.png"
will output a photo.png which will be the inverted image of photo.jpeg

**Comment by jywarren** [comment]
oh awesome!

**Comment by tech4GT** [comment]
@jywarren i think we can close this one now, thanks😁

**Comment by jywarren** [comment]
Perfect, thanks.

**Comment by jywarren** [comment]
Should we create an issue for `-o` being used to specify an output filename?

`-o  | --output [PATH]    | Directory where output will be stored. (optional)`

**Comment by tech4GT** [comment]
@jywarren we can do that but i would be doing that in the summer anyway
